File: core/normalizer.py
import re
from typing import Tuple, Optional, Dict, Any

# Regex patterns for matching numbers and units
DAYS_PATTERNS = [re.compile(r"(\d+)\s*d(ay)?s?", re.IGNORECASE)]
MONTHS_PATTERNS = [re.compile(r"(\d+)\s*m(onth)?s?", re.IGNORECASE)]
YEARS_PATTERNS = [re.compile(r"(\d+)\s*y(ear)?s?", re.IGNORECASE)]

def parse_tenure(tenure_str: str) -> Tuple[Optional[float], Optional[float], Optional[float]]:
    """
    Parses a tenure string (e.g. '7 Days', '1 Year 6 Months', '12 to 24 Months')
    and returns a tuple of (days, months, years).
    
    If it is a range (e.g., '46 days to 179 days'), we parse the lower bound as the starting point.
    """
    if not tenure_str:
        return None, None, None

    # Clean the string
    cleaned = tenure_str.strip()
    
    if not cleaned:
        return None, None, None
        
    if cleaned.isdigit():
        if int(cleaned) < 15:
            return None, None, None
    else:
        if not any(c.isdigit() for c in cleaned):
            return None, None, None
        if len(cleaned) < 3:
            has_digit = any(c.isdigit() for c in cleaned)
            has_unit = any(u in cleaned.lower() for u in ['y', 'm', 'd', 'w'])
            if not (has_digit and has_unit):
                return None, None, None

    # Exclude typical note indicator prefixes if they are at the very beginning of the string
    # e.g., "Less than 1 year" -> "1 year", "Up to 15 days" -> "15 days"
    cleaned_tmp = re.sub(r'^(?:less\s+than|below|up\s+to|under)\s+', '', cleaned, flags=re.IGNORECASE)

    # If it's a range like "12 to 24 Months", extract the first range component
    # Let's split on common range delimiters: "to", "-"
    # Example: "1 Year to 15 Months" -> Split to "1 Year" and "15 Months"
    parts = re.split(r'\b(?:to|less\s+than|-)\b', cleaned_tmp, flags=re.IGNORECASE)
    primary_part = parts[0].strip()

    total_days = 0.0
    total_months = 0.0
    total_years = 0.0

    found_any = False

    # Extract years
    for pattern in YEARS_PATTERNS:
        match = pattern.search(primary_part)
        if match:
            years = float(match.group(1))
            total_years += years
            total_months += years * 12.0
            total_days += years * 365.0
            found_any = True
            # Remove from string to prevent double matching
            primary_part = pattern.sub("", primary_part)

    # Extract months
    for pattern in MONTHS_PATTERNS:
        match = pattern.search(primary_part)
        if match:
            months = float(match.group(1))
            total_months += months
            total_days += months * 30.417  # Average days in month
            total_years += months / 12.0
            found_any = True
            primary_part = pattern.sub("", primary_part)

    # Extract days
    for pattern in DAYS_PATTERNS:
        match = pattern.search(primary_part)
        if match:
            days = float(match.group(1))
            total_days += days
            total_months += days / 30.417
            total_years += days / 365.0
            found_any = True
            primary_part = pattern.sub("", primary_part)

    # If no units were found but we have a raw number (e.g. "399"), assume it is days
    if not found_any:
        digits = re.search(r'^(\d+)$', primary_part)
        if digits:
            days = float(digits.group(1))
            total_days = days
            total_months = days / 30.417
            total_years = days / 365.0
            found_any = True

    if not found_any:
        return None, None, None

    return round(total_days, 1), round(total_months, 1), round(total_years, 2)


def normalize_rate(rate_str: str) -> Optional[float]:
    """
    Extracts a numeric float value from an interest rate string.
    Example: '7.10%' -> 7.10
    Example: '6.50 p.a.' -> 6.50
    """
    if not rate_str:
        return None

    # Replace commas, percentage symbols, and spaces
    cleaned = rate_str.replace("%", "").replace(",", "").strip()
    
    # Try finding the first decimal number
    match = re.search(r"(\d+\.?\d*)", cleaned)
    if match:
        try:
            return float(match.group(1))
        except ValueError:
            return None
    return None


def parse_tenure_to_days(bound_str: str) -> Optional[int]:
    """Helper to convert a single tenure block string to integer days."""
    cleaned = bound_str.strip()
    if not cleaned:
        return None
        
    total_days = 0.0
    found_any = False
    
    # Extract years
    for pattern in YEARS_PATTERNS:
        match = pattern.search(cleaned)
        if match:
            years = float(match.group(1))
            total_days += years * 365.0
            found_any = True
            cleaned = pattern.sub("", cleaned)
            
    # Extract months
    for pattern in MONTHS_PATTERNS:
        match = pattern.search(cleaned)
        if match:
            months = float(match.group(1))
            total_days += months * 30.417
            found_any = True
            cleaned = pattern.sub("", cleaned)
            
    # Extract days
    for pattern in DAYS_PATTERNS:
        match = pattern.search(cleaned)
        if match:
            days = float(match.group(1))
            total_days += days
            found_any = True
            cleaned = pattern.sub("", cleaned)
            
    if not found_any:
        digits = re.search(r'^(\d+)$', cleaned.strip())
        if digits:
            total_days = float(digits.group(1))
            found_any = True
            
    if found_any:
        return int(round(total_days))
    return None


def parse_tenure_range(tenure_str: str) -> Tuple[Optional[int], Optional[int]]:
    """Parses a tenure range to determine min_days and max_days bounds."""
    if not tenure_str:
        return None, None
        
    cleaned = tenure_str.strip()
    cleaned_tmp = re.sub(r'^(?:less\s+than|below|up\s+to|under)\s+', '', cleaned, flags=re.IGNORECASE)
    
    # Check for "and above" or "+"
    if re.search(r'\band\s+above\b|\b&\s+above\b|\b\+\b', cleaned, re.IGNORECASE):
        base_days = parse_tenure_to_days(cleaned_tmp)
        if base_days is not None:
            return base_days, 36500  # Cap at 100 years
            
    # Split range delimiters
    parts = re.split(r'\b(?:to|less\s+than|-)\b|(?:\bbut\s+<\b)|[<>-]', cleaned_tmp, flags=re.IGNORECASE)
    parts = [p.strip() for p in parts if p.strip()]
    
    if len(parts) >= 2:
        part1 = parts[0]
        part2 = parts[1]
        
        days2 = parse_tenure_to_days(part2)
        days1 = parse_tenure_to_days(part1)
        
        if days1 is not None and days2 is not None:
            p1_lower = part1.lower()
            p2_lower = part2.lower()
            has_unit1 = any(u in p1_lower for u in ["day", "month", "year", "week", "yr", "mth"]) or \
                        any(re.search(rf'\b{u}\b', p1_lower) for u in ["d", "m", "y", "w"])
            if not has_unit1:
                # Inherit unit of part2
                if "year" in p2_lower or "yr" in p2_lower or re.search(r'\by\b', p2_lower):
                    # Only inherit year if the raw number is small (e.g. < 12)
                    if days1 < 12:
                        days1 = days1 * 365
                elif "month" in p2_lower or "mth" in p2_lower or re.search(r'\bm\b', p2_lower):
                    # Only inherit month if the raw number is small (e.g. < 120)
                    if days1 < 120:
                        days1 = int(round(days1 * 30.417))
            return days1, days2
            
    elif len(parts) == 1:
        days = parse_tenure_to_days(parts[0])
        if days is not None:
            return days, days
            
    return None, None


def classify_fd_product(section_name: str, table_name: str, tenure_raw: str) -> Dict[str, Any]:
    """Classifies an FD product mapping its properties from table headings and tenure patterns."""
    sec_lower = section_name.lower() if section_name else ""
    tbl_lower = table_name.lower() if table_name else ""
    ten_lower = tenure_raw.lower() if tenure_raw else ""
    
    # 1. product_type
    bulk_keywords = [
        "bulk", "above 2 crore", "above 2cr", "above rs. 2 crore", "above rs 2 crore",
        "above rs. 1 crore", "above 1 crore", "3 cr. to 10 cr.", "above rs 5 crore",
        "above rs. 5 crore", "5 cr", "10 cr", "25 cr", "50 cr", "100 cr", "200 cr", "500 cr", "1000 cr",
        "above 2.00 crore", "above 1.00 crore", "above 5.00 crore"
    ]
    is_bulk = False
    for bk in bulk_keywords:
        if bk in sec_lower or bk in tbl_lower:
            is_bulk = True
            break
            
    product_type = "bulk_fd" if is_bulk else "retail_fd"
    
    # 2. deposit_category
    deposit_category = "regular"
    if "tax saver" in sec_lower or "tax saver" in tbl_lower or "tax-saver" in sec_lower or "tax-saver" in tbl_lower or "80c" in sec_lower or "80c" in tbl_lower:
        deposit_category = "tax_saver"
    elif "green" in sec_lower or "green" in tbl_lower or "eco" in sec_lower or "eco" in tbl_lower:
        deposit_category = "green_deposit"
        
    # 3. customer_segment
    has_resident = any(k in sec_lower or k in tbl_lower for k in ["domestic", "resident"])
    has_nre = "nre" in sec_lower or "nre" in tbl_lower
    has_nro = "nro" in sec_lower or "nro" in tbl_lower
    has_fcnr = "fcnr" in sec_lower or "fcnr" in tbl_lower
    
    matched_segments = []
    if has_resident:
        matched_segments.append("resident")
    if has_nre:
        matched_segments.append("nre")
    if has_nro:
        matched_segments.append("nro")
    if has_fcnr:
        matched_segments.append("fcnr")
        
    if len(matched_segments) > 1:
        customer_segment = "mixed"
    elif len(matched_segments) == 1:
        customer_segment = matched_segments[0]
    else:
        if "senior citizen special" in sec_lower or "senior citizen special" in tbl_lower or "senior special" in sec_lower or "senior special" in tbl_lower:
            customer_segment = "senior_citizen"
        else:
            customer_segment = "resident"
        
    # 4. callable
    callable_val = True
    non_callable_keywords = ["non-callable", "non callable", "without premature", "no premature", "uttam"]
    if any(k in sec_lower or k in tbl_lower for k in non_callable_keywords):
        callable_val = False
        
    # 5. scheme_type & scheme_name
    scheme_type = "regular_fd"
    scheme_name = None
    
    is_range = "to" in ten_lower or "-" in ten_lower or "less" in ten_lower or "below" in ten_lower or "above" in ten_lower or "or more" in ten_lower or "<" in ten_lower or ">" in ten_lower
    
    is_special = False
    if "green" in sec_lower or "green" in tbl_lower:
        is_special = True
        scheme_name = "Green Deposit"
    elif deposit_category == "tax_saver" or "tax saver" in ten_lower or "tax-saver" in ten_lower:
        is_special = True
        scheme_type = "tax_saver_fd"
        scheme_name = "Tax Saver FD"
    elif not is_range and any(k in ten_lower for k in ["day", "month", "year"]):
        cleaned_ten = re.sub(r'[*#$\s]+$', '', ten_lower).strip()
        if cleaned_ten not in ["1 year", "2 years", "3 years", "4 years", "5 years", "6 years", "7 years", "8 years", "9 years", "10 years", "7 days", "15 days", "30 days", "45 days", "90 days", "180 days"]:
            is_special = True
            scheme_name = tenure_raw.strip() + " FD"
            
    if is_special and scheme_type == "regular_fd":
        scheme_type = "special_fd"
        
    return {
        "product_type": product_type,
        "deposit_category": deposit_category,
        "customer_segment": customer_segment,
        "callable": callable_val,
        "scheme_type": scheme_type,
        "scheme_name": scheme_name
    }


def normalize_date_string(date_str: str) -> Optional[str]:
    """Normalizes a raw date string to YYYY-MM-DD format."""
    if not date_str:
        return None
    import datetime
    
    # Clean the string
    cleaned = date_str.strip()
    # Remove ordinal suffixes like 1st, 2nd, 3rd, 4th
    cleaned = re.sub(r'\b(\d+)(st|nd|rd|th)\b', r'\1', cleaned, flags=re.IGNORECASE)
    # Remove trailing/leading special characters
    cleaned = re.sub(r'[*#$\s]+$', '', cleaned)
    cleaned = re.sub(r'^\s*[:\-\/\.\s]+', '', cleaned)
    cleaned = re.sub(r'\s+', ' ', cleaned)
    
    # 1. Try DD.MM.YYYY or DD/MM/YYYY or DD-MM-YYYY
    match = re.search(r'\b(\d{1,2})[\.\-\/](\d{1,2})[\.\-\/](\d{4})\b', cleaned)
    if match:
        d, m, y = match.groups()
        try:
            dt = datetime.date(int(y), int(m), int(d))
            return dt.isoformat()
        except ValueError:
            pass
            
    # 2. Text month formats: e.g. "June 9, 2026", "1 June 2026", "June 2026"
    months_map = {
        "jan": 1, "feb": 2, "mar": 3, "apr": 4, "may": 5, "jun": 6,
        "jul": 7, "aug": 8, "sep": 9, "oct": 10, "nov": 11, "dec": 12,
        "january": 1, "february": 2, "march": 3, "april": 4, "june": 6,
        "july": 7, "august": 8, "september": 9, "october": 10, "november": 11, "december": 12
    }
    
    month_num = None
    for m_name, m_val in months_map.items():
        if re.search(r'\b' + m_name + r'\b', cleaned.lower()):
            month_num = m_val
            break
            
    if month_num is not None:
        # Try to find year
        year_match = re.search(r'\b(20\d{2})\b', cleaned)
        year = int(year_match.group(1)) if year_match else 2026  # Default to current year 2026
        
        # Try to find day
        numbers = re.findall(r'\b(\d{1,2})\b', cleaned)
        day = 1
        for num in numbers:
            val = int(num)
            if val != year and 1 <= val <= 31:
                day = val
                break
        try:
            dt = datetime.date(year, month_num, day)
            return dt.isoformat()
        except ValueError:
            pass
            
    return None


def extract_effective_date(text: str) -> Optional[str]:
    """Attempts to match an effective date pattern from page headings or footnotes."""
    if not text:
        return None
    match = re.search(r'\bw\.e\.f\.?\s*(?:from|date)?\s*([A-Za-z0-9\.\-\/\,\s]{6,25})', text, re.IGNORECASE)
    if match:
        date_str = match.group(1).strip()
        return normalize_date_string(date_str)
    
    # Try direct date formats in the text as a backup
    direct_normalized = normalize_date_string(text)
    if direct_normalized:
        return direct_normalized
        
    return None

File: core/normalizer.js
const DAYS_PATTERNS = [/(\d+)\s*d(ay)?s?/i];
const MONTHS_PATTERNS = [/(\d+)\s*m(onth)?s?/i];
const YEARS_PATTERNS = [/(\d+)\s*y(ear)?s?/i];

export function parseTenure(tenureStr) {
  if (!tenureStr) return [null, null, null];

  let cleaned = tenureStr.trim();
  if (!cleaned) return [null, null, null];

  if (/^\d+$/.test(cleaned)) {
    if (parseInt(cleaned, 10) < 15) {
      return [null, null, null];
    }
  } else {
    if (!/\d/.test(cleaned)) {
      return [null, null, null];
    }
    if (cleaned.length < 3) {
      const hasDigit = /\d/.test(cleaned);
      const hasUnit = /[ymdw]/i.test(cleaned);
      if (!(hasDigit && hasUnit)) {
        return [null, null, null];
      }
    }
  }

  let cleanedTmp = cleaned.replace(/^(?:less\s+than|below|up\s+to|under)\s+/i, "");
  const parts = cleanedTmp.split(/\b(?:to|less\s+than|-)\b/i);
  let primaryPart = parts[0].trim();

  let totalDays = 0.0;
  let totalMonths = 0.0;
  let totalYears = 0.0;
  let foundAny = false;

  // Years
  for (const pattern of YEARS_PATTERNS) {
    const match = pattern.exec(primaryPart);
    if (match) {
      const years = parseFloat(match[1]);
      totalYears += years;
      totalMonths += years * 12.0;
      totalDays += years * 365.0;
      foundAny = true;
      primaryPart = primaryPart.replace(pattern, "");
    }
  }

  // Months
  for (const pattern of MONTHS_PATTERNS) {
    const match = pattern.exec(primaryPart);
    if (match) {
      const months = parseFloat(match[1]);
      totalMonths += months;
      totalDays += months * 30.417;
      totalYears += months / 12.0;
      foundAny = true;
      primaryPart = primaryPart.replace(pattern, "");
    }
  }

  // Days
  for (const pattern of DAYS_PATTERNS) {
    const match = pattern.exec(primaryPart);
    if (match) {
      const days = parseFloat(match[1]);
      totalDays += days;
      totalMonths += days / 30.417;
      totalYears += days / 365.0;
      foundAny = true;
      primaryPart = primaryPart.replace(pattern, "");
    }
  }

  if (!foundAny) {
    const digits = /^\d+$/.exec(primaryPart.trim());
    if (digits) {
      const days = parseFloat(digits[0]);
      totalDays = days;
      totalMonths = days / 30.417;
      totalYears = days / 365.0;
      foundAny = true;
    }
  }

  if (!foundAny) return [null, null, null];

  return [
    Math.round(totalDays * 10) / 10,
    Math.round(totalMonths * 10) / 10,
    Math.round(totalYears * 100) / 100
  ];
}

export function normalizeRate(rateStr) {
  if (!rateStr) return null;
  const cleaned = String(rateStr).replace(/%/g, "").replace(/,/g, "").trim();
  const match = /(\d+\.?\d*)/.exec(cleaned);
  if (match) {
    const val = parseFloat(match[1]);
    return isNaN(val) ? null : val;
  }
  return null;
}

export function parseTenureToDays(boundStr) {
  let cleaned = boundStr.trim();
  if (!cleaned) return null;

  let totalDays = 0.0;
  let foundAny = false;

  for (const pattern of YEARS_PATTERNS) {
    const match = pattern.exec(cleaned);
    if (match) {
      const years = parseFloat(match[1]);
      totalDays += years * 365.0;
      foundAny = true;
      cleaned = cleaned.replace(pattern, "");
    }
  }

  for (const pattern of MONTHS_PATTERNS) {
    const match = pattern.exec(cleaned);
    if (match) {
      const months = parseFloat(match[1]);
      totalDays += months * 30.417;
      foundAny = true;
      cleaned = cleaned.replace(pattern, "");
    }
  }

  for (const pattern of DAYS_PATTERNS) {
    const match = pattern.exec(cleaned);
    if (match) {
      const days = parseFloat(match[1]);
      totalDays += days;
      foundAny = true;
      cleaned = cleaned.replace(pattern, "");
    }
  }

  if (!foundAny) {
    const digits = /^\d+$/.exec(cleaned.trim());
    if (digits) {
      totalDays = parseFloat(digits[0]);
      foundAny = true;
    }
  }

  if (foundAny) return Math.round(totalDays);
  return null;
}

export function parseTenureRange(tenureStr) {
  if (!tenureStr) return [null, null];

  const cleaned = tenureStr.trim();
  const cleanedTmp = cleaned.replace(/^(?:less\s+than|below|up\s+to|under)\s+/i, "");

  if (/\band\s+above\b|\b&\s+above\b|\b\+\b/i.test(cleaned)) {
    const baseDays = parseTenureToDays(cleanedTmp);
    if (baseDays !== null) {
      return [baseDays, 36500]; // Cap at 100 years
    }
  }

  // Split range delimiters
  const parts = cleanedTmp.split(/\b(?:to|less\s+than|-)\b|(?:\bbut\s+<\b)|[<>-]/i)
                          .map(p => p.trim())
                          .filter(p => p);

  if (parts.length >= 2) {
    const part1 = parts[0];
    const part2 = parts[1];

    let days2 = parseTenureToDays(part2);
    let days1 = parseTenureToDays(part1);

    if (days1 !== null && days2 !== null) {
      const p1Lower = part1.toLowerCase();
      const p2Lower = part2.toLowerCase();

      const hasUnit1 = ["day", "month", "year", "week", "yr", "mth"].some(u => p1Lower.includes(u)) ||
                       [/\bd\b/, /\bm\b/, /\by\b/, /\bw\b/].some(rx => rx.test(p1Lower));

      if (!hasUnit1) {
        // Inherit unit of part2
        if (p2Lower.includes("year") || p2Lower.includes("yr") || /\by\b/.test(p2Lower)) {
          if (days1 < 12) {
            days1 = days1 * 365;
          }
        } else if (p2Lower.includes("month") || p2Lower.includes("mth") || /\bm\b/.test(p2Lower)) {
          if (days1 < 120) {
            days1 = Math.round(days1 * 30.417);
          }
        }
      }
      return [days1, days2];
    }
  } else if (parts.length === 1) {
    const days = parseTenureToDays(parts[0]);
    if (days !== null) {
      return [days, days];
    }
  }

  return [null, null];
}

export function classifyFDProduct(sectionName, tableName, tenureRaw) {
  const secLower = sectionName ? sectionName.toLowerCase() : "";
  const tblLower = tableName ? tableName.toLowerCase() : "";
  const tenLower = tenureRaw ? tenureRaw.toLowerCase() : "";

  // 1. product_type
  const bulkKeywords = [
    "bulk", "above 2 crore", "above 2cr", "above rs. 2 crore", "above rs' 2 crore",
    "above rs. 1 crore", "above 1 crore", "3 cr. to 10 cr.", "above rs 5 crore",
    "above rs. 5 crore", "5 cr", "10 cr", "25 cr", "50 cr", "100 cr", "200 cr", "500 cr", "1000 cr",
    "above 2.00 crore", "above 1.00 crore", "above 5.00 crore"
  ];
  const isBulk = bulkKeywords.some(bk => secLower.includes(bk) || tblLower.includes(bk));
  const productType = isBulk ? "bulk_fd" : "retail_fd";

  // 2. deposit_category
  let depositCategory = "regular";
  if (
    secLower.includes("tax saver") || tblLower.includes("tax saver") ||
    secLower.includes("tax-saver") || tblLower.includes("tax-saver") ||
    secLower.includes("80c") || tblLower.includes("80c") ||
    tenLower.includes("tax saver") || tenLower.includes("tax-saver")
  ) {
    depositCategory = "tax_saver";
  } else if (
    secLower.includes("green") || tblLower.includes("green") ||
    secLower.includes("eco") || tblLower.includes("eco")
  ) {
    depositCategory = "green_deposit";
  }

  // 3. customer_segment
  const hasResident = ["domestic", "resident"].some(k => secLower.includes(k) || tblLower.includes(k));
  const hasNre = secLower.includes("nre") || tblLower.includes("nre");
  const hasNro = secLower.includes("nro") || tblLower.includes("nro");
  const hasFcnr = secLower.includes("fcnr") || tblLower.includes("fcnr");

  const matchedSegments = [];
  if (hasResident) matchedSegments.push("resident");
  if (hasNre) matchedSegments.push("nre");
  if (hasNro) matchedSegments.push("nro");
  if (hasFcnr) matchedSegments.push("fcnr");

  let customerSegment = "resident";
  if (matchedSegments.length > 1) {
    customerSegment = "mixed";
  } else if (matchedSegments.length === 1) {
    customerSegment = matchedSegments[0];
  } else {
    if (
      secLower.includes("senior citizen special") || tblLower.includes("senior citizen special") ||
      secLower.includes("senior special") || tblLower.includes("senior special")
    ) {
      customerSegment = "senior_citizen";
    } else {
      customerSegment = "resident";
    }
  }

  // 4. callable
  let callableVal = true;
  const nonCallableKeywords = ["non-callable", "non callable", "without premature", "no premature", "uttam"];
  if (nonCallableKeywords.some(k => secLower.includes(k) || tblLower.includes(k))) {
    callableVal = false;
  }

  // 5. scheme_type & scheme_name
  let schemeType = "regular_fd";
  let schemeName = null;

  const isRange = ["to", "-", "less", "below", "above", "or more", "<", ">"].some(k => tenLower.includes(k));
  let isSpecial = false;

  if (secLower.includes("green") || tblLower.includes("green")) {
    isSpecial = true;
    schemeName = "Green Deposit";
  } else if (depositCategory === "tax_saver" || tenLower.includes("tax saver") || tenLower.includes("tax-saver")) {
    isSpecial = true;
    schemeType = "tax_saver_fd";
    schemeName = "Tax Saver FD";
  } else if (!isRange && ["day", "month", "year"].some(k => tenLower.includes(k))) {
    const cleanedTen = tenureRaw.replace(/[*#$\s]+$/, "").trim().toLowerCase();
    const standardTenures = [
      "1 year", "2 years", "3 years", "4 years", "5 years", "6 years", "7 years", "8 years", "9 years", "10 years",
      "7 days", "15 days", "30 days", "45 days", "90 days", "180 days"
    ];
    if (!standardTenures.includes(cleanedTen)) {
      isSpecial = true;
      schemeName = tenureRaw.trim() + " FD";
    }
  }

  if (isSpecial && schemeType === "regular_fd") {
    schemeType = "special_fd";
  }

  return {
    product_type: productType,
    deposit_category: depositCategory,
    customer_segment: customerSegment,
    callable: callableVal,
    scheme_type: schemeType,
    scheme_name: schemeName
  };
}

export function normalizeDateString(dateStr) {
  if (!dateStr) return null;

  let cleaned = dateStr.trim();
  cleaned = cleaned.replace(/\b(\d+)(st|nd|rd|th)\b/gi, "$1");
  cleaned = cleaned.replace(/[*#$\s]+$/, "");
  cleaned = cleaned.replace(/^\s*[:\-/\.\s]+/, "");
  cleaned = cleaned.replace(/\s+/g, " ");

  // 1. Try DD.MM.YYYY or DD/MM/YYYY or DD-MM-YYYY
  const match = /\b(\d{1,2})[\.\-\/](\d{1,2})[\.\-\/](\d{4})\b/.exec(cleaned);
  if (match) {
    const d = parseInt(match[1], 10);
    const m = parseInt(match[2], 10);
    const y = parseInt(match[3], 10);
    try {
      const date = new Date(Date.UTC(y, m - 1, d));
      if (date.getUTCFullYear() === y && date.getUTCMonth() === m - 1 && date.getUTCDate() === d) {
        return date.toISOString().split("T")[0];
      }
    } catch (e) {}
  }

  // 2. Text month formats: e.g. "June 9, 2026", "1 June 2026", "June 2026"
  const monthsMap = {
    jan: 1, feb: 2, mar: 3, apr: 4, may: 5, jun: 6,
    jul: 7, aug: 8, sep: 9, oct: 10, nov: 11, dec: 12,
    january: 1, february: 2, march: 3, april: 4, june: 6,
    july: 7, august: 8, september: 9, october: 10, november: 11, december: 12
  };

  let monthNum = null;
  const cleanedLower = cleaned.toLowerCase();
  for (const [mName, mVal] of Object.entries(monthsMap)) {
    const rx = new RegExp(`\\b${mName}\\b`);
    if (rx.test(cleanedLower)) {
      monthNum = mVal;
      break;
    }
  }

  if (monthNum !== null) {
    const yearMatch = /\b(20\d{2})\b/.exec(cleaned);
    const year = yearMatch ? parseInt(yearMatch[1], 10) : 2026;

    const numbers = [];
    const numRegex = /\b(\d{1,2})\b/g;
    let numMatch;
    while ((numMatch = numRegex.exec(cleaned)) !== null) {
      numbers.push(parseInt(numMatch[1], 10));
    }

    let day = 1;
    for (const num of numbers) {
      if (num !== year && num >= 1 && num <= 31) {
        day = num;
        break;
      }
    }

    try {
      const date = new Date(Date.UTC(year, monthNum - 1, day));
      if (date.getUTCFullYear() === year && date.getUTCMonth() === monthNum - 1 && date.getUTCDate() === day) {
        return date.toISOString().split("T")[0];
      }
    } catch (e) {}
  }

  return null;
}

export function extractEffectiveDate(text) {
  if (!text) return null;
  const match = /\bw\.e\.f\.?\s*(?:from|date)?\s*([A-Za-z0-9\.\-\/\,\s]{6,25})/i.exec(text);
  if (match) {
    const dateStr = match[1].replace(/[*#$\s]+$/, "").trim();
    return normalizeDateString(dateStr);
  }

  const directNormalized = normalizeDateString(text);
  if (directNormalized) {
    return directNormalized;
  }

  return null;
}
